feat: derive from Code enum

vmx opened the pull request:
Instead of deriving a Multihash enum, derive from a Code enum.

This commit also combines the previously generated `Multihash` enum with
the `RawMultihash` struct. There is now a single `Multihash` type which
is generic over its digest size.

The derived `Code` enum is used for hashing, the `Multihash` only contains
static, general data.

Closes #12, #13.

There a still a few TODOs, but I think it's ready for a review, to see if we want to
do things that way. Things now work more the way they used to work.

vmx committed:
feat: derive from Code enum

Instead of deriving a Multihash enum, derive from a Code enum.

This commit also combines the previously generated `Multihash` enum with
the `RawMultihash` struct. There is now a single `Multihash` type which
is generic over its digest size.

The derived `Code` enum is used for hashing, the `Multihash` only contains
static, general data.

Closes #12, #13.

vmx committed:
fix: don't clone the digest

dvc94ch commented:
Why make it an attribute macro?

```rust
#[derive(Multihash)]
#[mh(max_size = U64)]
pub enum Multihash {
    ...
}
```
Does max_size have a default? What if I lie about the max_size? Is the max_size checked?

You removed the MultihashDigest trait, without a replacement. That seems very fishy. You presumably need a `MultihashCode` trait.

~I'd like to move this stuff into rust-ipld, as it's impossible to tell how api design changes will affect the rest of the code.~ Looking good, after rereading it I see how it works.

Also please fix the scale/serde tests.

dvc94ch reviewed (changes requested):
- [x] add a trait

```rust
pub trait MultihashCode: From<u64> + TryInto<u64> + Copy + Send + Sync + Unpin + 'static {
    type MaxSize: ArrayLength<u8>;
    fn digest(&self, &[u8]) -> Multihash<Self::MaxSize>;
}
```

- [x] fix scale/serde tests
- [x] use a derive macro instead of an attribute macro

```rust
#[derive(Multihash)]
#[mh(max_size = U64)]
pub enum Multihash {
    ...
}
```

vmx commented:
> Does max_size have a default? What if I lie about the max_size? Is the max_size checked?

If you lie about it, then your code will panic. I don't want to have an implicit default, so that people need to set it. I want to expand the documentation, so that people know about what this is about. I had hoped I could automatically derive the size from the digests being used, but as far as I can tell it's not possible at compile time, as macros don't have access to type information.

I know it's not great, but I think it's an OK trade-off between flexibility and usability.


> You removed the MultihashDigest trait, without a replacement. That seems very fishy.

It was no longer needed as the is only a single `Multihash` implementation anymore.


> You presumably need a `MultihashCode` trait.
> ```rust
> pub trait MultihashCode: From<u64> + TryInto<u64> + Copy + Send + Sync + Unpin + 'static {
>     type MaxSize: ArrayLength<u8>;
>     fn digest(&self, &[u8]) -> Multihash<Self::MaxSize>;
> }
> ```

I haven't yet tried to integrate my version into rust-cid or into rust-ipld. Is that where the requirement for such a trait comes from? I'd say less traits are always favourable.

> * [x]  fix scale/serde tests

Yes, that's on the TODO list. I might need help with that (I'll let you know if I can't get it working).

> * use a derive macro instead of an attribute macro
> #[derive(Multihash)]
#[mh(max_size = U64)]
pub enum Multihash {
    ...
}

Oh wow, I didn't know I can also use the `#[mh()]` attribute at the enum itself. The only reason to use an attribute macro was that I needed that `max_size` parameter.

dvc94ch commented:
> I haven't yet tried to integrate my version into rust-cid or into rust-ipld. Is that where the requirement for such a trait comes from? I'd say less traits are always favourable.

It's how rust works. If you want to make a component replaceable, you need to define it's api in a trait and parameterize types over that trait. How do you expect custom code tables to be used?

vmx commented:
> How do you expect custom code tables to be used?

I expect that you use `u64` for multihash codes code and then do a `MyCodeTable::from(u64_code)`-type conversions. Though I can see the point that you might want to know the maximum size of a code table.

I would add this trait if it turns out it is needed, but I'm also not against it. So if you think it makes sense, I'll just add it (I was just interested if there are already use cases I've missed).

dvc94ch commented:
bitswap and ipld needs to be to construct a multihash. Do you prefer we add traits to bitswap and libipld? That way they'd be agnostic over the multihash implementation.

dvc94ch commented:
Something like this is required for both ipld and bitswap:
```rust
fn needs_to_create_a_hash<M: MultihashCode>(code: u64, data: &[u8]) -> Multihash<M::MaxSize> {
    M::from(code).digest(data)
}
```

vmx commented:
> bitswap and ipld needs to be to construct a multihash. Do you prefer we add traits to bitswap and libipld? That way they'd be agnostic over the multihash implementation.

If there is already two cased where a trait like this is needed, it's a sign for me that it should be added in a single place. I'll add the trait here and then I'll see how it is used in bitswap and ipld, to get a better understanding for future cases.

vmx commented:
> ```rust
> fn needs_to_create_a_hash<M: MultihashCode>(code: u64, data: &[u8]) -> Multihash<M::MaxSize> {
>     M::from(code).digest(data)
> }
> ```

I'm sorry, that's super obvious now :) You of course don't want to have a specific code table hard-coded there.

Thanks for taking the time explaining it to me.

dvc94ch commented:
I think we can infer the max size, since a digest has the form `Wrapper<U**>`. So you can parse the generic parameter of the digest.

dvc94ch commented:
Event better would actually be to use https://doc.rust-lang.org/std/mem/fn.size_of.html (we need const generics for that :/, so parsing the type name is the only solution for now)

vmx committed:
feat: introduce `max_size` attribute

The digests are stack allocated. We cannot determine the biggest digest size
that is used at compile time, hence you need to specify the maximum allocated
size your custom code table should use via `[#mh(max_size = …)]` on the enum
you derive the table from.

vmx commented:
This is still WIP. This is a rebased version of the already reviewed code. The only major change is that it now uses a derived macro instead of an attribute macro. There are still things left to do.

vmx committed:
fix: enable the arb module again

vmx committed:
fix: properly use local versions of imports

This way you only need to import the derive and not also additional
(seemingly unrelated) types.

vmx committed:
fix: make Serde support work again

vmx committed:
fix: make Parity Scale Codec support work again

vmx commented:
The only missing piece is adding the trait for `Code`.

vmx committed:
feat: create MultihashCode trait

Instead of just implementing it on the code enum, use a trait instead.
This way other libraries can create digests without knowing the concrete
type.

vmx commented:
@dvc94ch it's ready for review!

dvc94ch commented:
Looking good. Can't we parse max size by parsing the generic type argument?

dvc94ch commented on src/multihash.rs:
```diff
@@ -1,121 +1,99 @@
-pub trait MultihashDigest: Clone + Debug + Eq + Send + Sync + 'static {
-    /// Returns the hash of the input.
-    fn new(code: u64, input: &[u8]) -> Result<Self, Error>;
+pub trait MultihashCode: TryFrom<u64> + Into<u64> + Send + Sync + 'static {
```
Can you make it `Unpin`, `Copy`, `Eq` and `Debug`? Those can be easily derived for enums and make life easier.

vmx commented:
> Can't we parse max size by parsing the generic type argument?

I thought about it and looked into it as I also would've loved to have it done automatically. The problem is that you don't have type information available in macros. So it would be error prone even when we would found a way (which I haven't found anyway).

vmx committed:
fix: make MultihashCode also need Unpin + Copy + Eq + Debug

This should make the MultihashCode easier to use.

dvc94ch commented:
It's a syn::Path, so you need to skip to the syn::Type, then get the generic argument which is a syn::Path, get the syn::Type and then find the largest, via lexicological sorting and use that

vmx commented:
> It's a syn::Path, so you need to skip to the syn::Type, then get the generic argument which is a syn::Path, get the syn::Type and then find the largest, via lexicological sorting and use that

It won't work with type aliases. I'm not sure if it's a problem or not. I don't think people will use aliases, but if, then it's really a hard to find problem. I also think setting the `max_size` gives a good indication on how things actually work, so it's less magic an more explicit.

dvc94ch commented:
I guess you could return an error in that case, so it wouldn't be a problem. I don't even mind it being explicit, especially as it's a breaking change to increase the size. I'm more concerned about people making mistakes. Maybe we should return a warning at least?

vmx commented:
> I'm more concerned about people making mistakes. Maybe we should return a warning at least?

I thought the same. It's not a problem when you've specified things and never change anything. But I can see the problem that someone adds another, bigger hash digest and then suddenly things are breaking randomly.

What about emitting a warning following this algorithm:

 1. Get all the digest sizes
 2. If one type is not `U<number>` just return, do nothing else (it could be an indicator that someone uses an alias)
 3. If the biggest type is bigger than `max_size`, print a warning (I hope I can make the string into numbers and compare them)

dvc94ch commented:
I'd tend towards 2. being a warning and 3. an error, but it sounds like a plan.

vmx commented:
Would it be OK if I do this on a separate PR and we merge this one? As this is a major change, I'd like to get this one merged sooner rather than later, so that no one bases it's own work on old stuff (in case people work on PRs).
